core/drivers/crypto/caam/acipher/caam_dsa.c

```diff
@@ -0,0 +1,778 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2019-2021 NXP
+ *
+ * Implementation of DSA functions
+ */
+#include <caam_acipher.h>
+#include <caam_common.h>
+#include <caam_hal_ctrl.h>
+#include <caam_jr.h>
+#include <caam_utils_mem.h>
+#include <caam_utils_status.h>
+#include <drvcrypt.h>
+#include <drvcrypt_acipher.h>
+#include <mm/core_memprot.h>
+#include <tee/cache.h>
+#include <string.h>
+
+#include "local.h"
+
+#ifdef CFG_CAAM_64BIT
+#define MAX_DESC_KEY_GEN 14
+#define MAX_DESC_SIGN    19
+#define MAX_DESC_VERIFY  21
+#else
+#define MAX_DESC_KEY_GEN 9
+#define MAX_DESC_SIGN    12
+#define MAX_DESC_VERIFY  13
+#endif
+
+/*
+ * Definition of the local DSA Keypair
+ *   Domain Parameters (p, q, g)
+ *   Private Key format (x)
+ *   Public Key format (y)
+ */
+struct caam_dsa_keypair {
+	struct caambuf g; /* Generator */
+	struct caambuf p; /* Prime Number (L bits) */
+	struct caambuf q; /* Subprime Number (N bits) */
+	struct caambuf x; /* Private key */
+	struct caambuf y; /* Public key */
+};
+
+/*
+ * Free local DSA keypair
+ *
+ * @key  DSA keypair
+ */
+static void do_keypair_free(struct caam_dsa_keypair *key)
+{
+	caam_free_buf(&key->g);
+	caam_free_buf(&key->p);
+	caam_free_buf(&key->q);
+	caam_free_buf(&key->x);
+	caam_free_buf(&key->y);
+}
+
+/*
+ * If all DSA parameters p, q and g are present, convert them from bignumbers
+ * to local buffers (via keypair object). Otherwise generate them.
+ *
+ * @outkey    [out] Output keypair in local format
+ * @key       Input key in TEE Crypto format
+ * @l_bytes   Prime p size in bytes
+ * @n_bytes   Subprime q size in bytes
+ */
+static TEE_Result get_keypair_domain_params(struct caam_dsa_keypair *outkey,
+					    const struct dsa_keypair *key,
+					    size_t l_bytes, size_t n_bytes)
+{
+	enum caam_status retstatus = CAAM_OUT_MEMORY;
+	size_t in_q_size = 0;
+	size_t in_p_size = 0;
+	size_t in_g_size = 0;
+	struct prime_data_dsa prime = { };
+
+	DSA_TRACE("DSA conv key param (p, g) of %zu bytes and (q) of %zu bytes",
+		  l_bytes, n_bytes);
+
+	retstatus = caam_calloc_buf(&outkey->q, n_bytes);
+	if (retstatus != CAAM_NO_ERROR)
+		return caam_status_to_tee_result(retstatus);
+
+	retstatus = caam_calloc_buf(&outkey->g, l_bytes);
+	if (retstatus != CAAM_NO_ERROR)
+		return caam_status_to_tee_result(retstatus);
+
+	retstatus = caam_calloc_buf(&outkey->p, l_bytes);
+	if (retstatus != CAAM_NO_ERROR)
+		return caam_status_to_tee_result(retstatus);
+
+	/*
+	 * Get all inputs parameters size, if one of them is not
+	 * define generate new parameters
+	 */
+	in_g_size = crypto_bignum_num_bytes(key->g);
+	in_p_size = crypto_bignum_num_bytes(key->p);
+	in_q_size = crypto_bignum_num_bytes(key->q);
+
+	if (!in_q_size || !in_g_size || !in_p_size) {
+		/* Generate DSA parameters: Generator G and Primes P/Q */
+		prime.g = &outkey->g;
+		prime.p = &outkey->p;
+		prime.q = &outkey->q;
+
+		retstatus = caam_prime_dsa_gen(&prime);
+		DSA_TRACE("Generate G and Primes P/Q returned 0x%" PRIx32,
```

Comment:
restatus is an `enum caam_status` so `int` here, not a 32bit. s/0x%" PRIx32/%#x/

Reply:
Ok